code/modules/projectiles/sentries.dm

```diff
@@ -0,0 +1,421 @@
+	return ..()
+	
+//-----------------------------------------------------------------
+// Interaction
+
+/obj/machinery/deployable/mounted/sentry/attackby(obj/item/I, mob/user, params)
+	var/obj/item/weapon/gun/gun = internal_item
+	if(istype(I, gun.sentry_battery_type))
+		internal_item.attackby(I, user, params)
+		return
+	return ..()
+
+/obj/machinery/deployable/mounted/sentry/AltClick(mob/user)
+	. = ..()
+	internal_item.AltClick(user)
+
+/obj/machinery/deployable/mounted/sentry/on_set_interaction(mob/user)
+	. = ..()
+	to_chat(user, "<span class='notice'>You disable the [src]'s automatic to operate it manually.</span>")
+	set_on(FALSE)
+
+/obj/machinery/deployable/mounted/sentry/on_unset_interaction(mob/user)
+	. = ..()
+	to_chat(user, "<span class='notice'>You stop using the [src] and its automatic functions re-activate</span>")
+	set_on(TRUE)
+
+/obj/machinery/deployable/mounted/sentry/attack_hand(mob/living/user)
+	. = ..()
+	if(!. || !CHECK_BITFIELD(machine_stat, KNOCKED_DOWN))
+		return
+	user.visible_message("<span class='notice'>[user] begins to set [src] upright.</span>",
+	"<span class='notice'>You begin to set [src] upright.</span>")
+
+	if(!do_after(user, 2 SECONDS, TRUE, src, BUSY_ICON_BUILD))
+		return
+
+	user.visible_message("<span class='notice'>[user] sets [src] upright.</span>",
+	"<span class='notice'>You set [src] upright.</span>")
+
+	DISABLE_BITFIELD(machine_stat, KNOCKED_DOWN)
+	density = TRUE
+	set_on(TRUE)
+
+/obj/machinery/deployable/mounted/sentry/interact(mob/user, manual_mode = FALSE)
+	var/obj/item/weapon/gun/gun = internal_item
+	if(manual_mode)
+		return ..()
+
+	if(CHECK_BITFIELD(machine_stat, KNOCKED_DOWN))
+		return TRUE
+	
+	if(CHECK_BITFIELD(gun.turret_flags, TURRET_IMMOBILE))
+		to_chat(user, "<span class='warning'>[src]'s panel is completely locked, you can't do anything.</span>")
+		return TRUE
+
+	ui_interact(user)
+
+/obj/machinery/deployable/mounted/sentry/ui_interact(mob/user, datum/tgui/ui)
+
+	if(CHECK_BITFIELD(machine_stat, KNOCKED_DOWN))
+		return
+
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Sentry", "Sentry Gun")
+		ui.open()
+
+/obj/machinery/deployable/mounted/sentry/ui_data(mob/user)
+	var/obj/item/weapon/gun/gun = internal_item
+	. = list(
+		"name" = copytext(name, 2),
+		"rounds" = (gun.current_mag ? gun.current_mag.current_rounds : 0),
+		"rounds_max" = (gun.current_mag ? gun.current_mag.max_rounds : gun.max_shells),
+		"fire_mode" = gun.gun_firemode,
+		"health" = obj_integrity,
+		"health_max" = max_integrity,
+		"has_cell" = (gun.sentry_battery ? 1 : 0),
+		"cell_charge" = gun.sentry_battery ? gun.sentry_battery.charge : 0,
+		"cell_maxcharge" = gun.sentry_battery ? gun.sentry_battery.maxcharge : 0,
+		"safety_toggle" = CHECK_BITFIELD(gun.turret_flags, TURRET_SAFETY),
+		"manual_override" = operator,
+		"alerts_on" = CHECK_BITFIELD(gun.turret_flags, TURRET_ALERTS),
+		"radial_mode" = CHECK_BITFIELD(gun.turret_flags, TURRET_RADIAL),
+	)
```

Comment:
the ones that are not updated as often should be put on ui_static_data() and updated accordingly